zebra-consensus/src/checkpoint.rs

```diff
@@ -35,7 +45,41 @@ struct CheckpointVerifier<S> {
     /// which only happen in the last few hundred blocks in the chain.
     /// (zcashd allows chain reorganizations up to 99 blocks, and prunes
     /// orphaned side-chains after 288 blocks.)
+    ///
+    /// Checkpoints must start at zero, and have equal spacing. There must
+    /// be at least one checkpoint, the genesis block.
     checkpoint_list: Arc<HashMap<BlockHeight, BlockHeaderHash>>,
+
+    // Derived Variables
+    //
+    /// The height of the maximum checkpoint.
+    ///
+    /// Can be zero, if there is a single genesis block checkpoint.
+    max_checkpoint_height: BlockHeight,
+
+    /// The fixed spacing between checkpoints.
+    ///
+    /// Can be zero, if there is a single genesis block checkpoint.
+    checkpoint_frequency: BlockHeight,
```

Comment:
The reason is so that the `CheckpointInterval`s can be indexed by `block_height % checkpoint_frequency`, and this is used as the key for the `queued` hashmap.

Reply:
As @yaahc suggested, I think we should use `BTreeMap<...>` as the data store. That way, we can call `range(Excluded(current_checkpoint_height), ...)` to get the next checkpoint.

Using `BTreeMap`, we can accommodate checkpoints that aren't a fixed stride apart. We have two mandatory checkpoints so far, genesis and sapling:
```
main net: 0, 419200
test net: 0, 280000
```
https://zips.z.cash/zip-0205

The main net sapling checkpoint makes it impossible to use a 500-block checkpoint stride, because 419200 is not a multiple of 500. If we wanted to use a common stride across networks, our options would be: 400, 800, or 1600.

But I'd rather remove the constraint entirely. We might end up with another mandatory checkpoint in future, or find it useful to have arbitrary checkpoints during testing or debugging. (Checkpoints are also a useful way of doing a quick validation failure hotfix, while we work on fixing the underlying issue.)

Reply:
The Sapling checkpoint is only mandatory in the sense that we must checkpoint *up to* Sapling activation -- there's no requirement that we checkpoint on that block exactly.  But removing the constraint entirely is nice.

Reply:
Removed the fixed stride requirement in 293e9d17c5672782e2b0adf50b74979fae417a14, by using BTreeMap.